Windows builds using AppVeyor

tiagocoutinho opened the pull request:
* based on NexeyaSGara work
* updates to tango 9.3
* add python 3.7, drop 3.4 and 3.5
* fix build on VC9 64bits
* add support for wheel

NexeyaSGara committed:
initial version

NexeyaSGara committed:
Create CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update device_proxy.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update pipe.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update device_impl.cpp

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update __init__.py

NexeyaSGara committed:
Fix version

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Fix indent

NexeyaSGara committed:
Fix new tango dll directory

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Change zmq to 4.0.5

NexeyaSGara committed:
Update tango package

NexeyaSGara committed:
Fix Tango version number

NexeyaSGara committed:
Use static only

NexeyaSGara committed:
Update tango version

NexeyaSGara committed:
Update Boost and tango

NexeyaSGara committed:
Update tango

NexeyaSGara committed:
Update Tango

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update Tango

NexeyaSGara committed:
Update tango

NexeyaSGara committed:
Force /MT

NexeyaSGara committed:
Update Tango

NexeyaSGara committed:
Add compile option /MT

NexeyaSGara committed:
Update Tango

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Update CMakeLists.txt

NexeyaSGara committed:
Add patch vor vc15

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Update device_attribute.cpp

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Update device_proxy.cpp

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Update api_util.cpp

NexeyaSGara committed:
Update attribute.cpp

NexeyaSGara committed:
Update attr.cpp

NexeyaSGara committed:
Update command.cpp

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Update dserver.cpp

NexeyaSGara committed:
Update log4tango.cpp

NexeyaSGara committed:
Update pipe.cpp

NexeyaSGara committed:
Update wattribute.cpp

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Add vc15 patch

NexeyaSGara committed:
Adding py36

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Fix msvcstub.lib path

NexeyaSGara committed:
Update appveyor.yml

NexeyaSGara committed:
Add Python version selector

NexeyaSGara committed:
Selector for link to right python version lib

NexeyaSGara committed:
Fix artifacts discovery

NexeyaSGara committed:
Fix boost download link

NexeyaSGara committed:
Fixing spaces

vxgmichel committed:
All executors use the same context detection

Commit:
Refactor and version fix

Commit:
forget this file on last commit

tiagocoutinho committed:
remove duplicate DEFINE

tiagocoutinho committed:
fix broken line

tiagocoutinho committed:
add missing include

tiagocoutinho committed:
Move missing tango C++ definitions to tgutils.h

Since tango C++ 9.3 there is no need for the missing definitions

tiagocoutinho committed:
Update CMakeLists

* update pytango version
* make sure we build specific python version and platform with correct MSCV
* drop python 3.4 and 3.5 builds and add python 3.7

tiagocoutinho committed:
cmake: separate Release directories for each build

tiagocoutinho committed:
cmake: fix error on VC9 x64: base_types.cpp needs bigobj size flag

tiagocoutinho committed:
winsetup: better build messages

tiagocoutinho committed:
winsetup: add support to build wheel

tiagocoutinho committed:
winsetup: fix bug with pypi platform name when uploading wheel files

ajoubertza commented on tango/__init__.py:
```diff
@@ -121,12 +121,18 @@ def __prepare_nt():
+        tango_dll_path = tango_dll_path.lower()
+        if os.path.exists(tango_dll_path) and \
+           tango_dll_path not in PATH.lower():
             os.environ['PATH'] += ";" + tango_dll_path
```
I think the check that it is in PATH variable already should be a separate condition inside this `if` block.  If the path exists, but isn't in the PATH variable already, then we won't add it - we just fall back to the the "Tango C++ could not be found".  That seems incorrect.  Rather:
```
        if os.path.exists(tango_dll_path):
            if tango_dll_path not in PATH.lower():
                os.environ['PATH'] += ";" + tango_dll_path
        else:
```

tiagocoutinho replied:
Better yet: we can get rid of the `__prepare_nt()` function completely.
This is an artefact from older versions of PyTango. It is completely useless now.

ajoubertza replied:
@tiagocoutinho This still needs to be addressed - either fix, or remove the whole method.

ajoubertza commented on CMakeLists.txt:
```diff
@@ -0,0 +1,254 @@
+#need to define the version of the library
+set(MAJOR_VERSION "9")
+set(MINOR_VERSION "3")
+set(PATCH_VERSION "1")
```
It would be better to derive this from the git tag - see comments in previous PR https://github.com/tango-controls/pytango/pull/176#discussion_r237493763

Alternatively, we can do it manually, but then [the docs](https://github.com/tango-controls/pytango/blob/develop/doc/how-to-contribute.rst#releasing-a-new-version) need to be updated - there are two section about "bump the version" that need to mention this file.

tiagocoutinho replied:
Agree.

@sdebionne you offered some help for PR #176. Could you give me a hand with this one?

sdebionne replied:
@tiagocoutinho Sure do! Today is a holiday in France, but I'll look at it tomorrow. If it cannot wait, look at `cmake/project_version.*` in [esrf-bliss/Lima](https://github.com/esrf-bliss/Lima) repo. You need to `include(cmake/project_version.cmake)` right after the project definition.

tiagocoutinho replied:
No hurry. Enjoy you holiday :-)

sdebionne replied:
@tiagocoutinho See the PR on your fork.

Couple of questions about the `CMakeLists.txt`:

 - Is there a reason to stick with CMake 3.0.2 (5 years old)?
 - Why not use `find_package()` to find Boost?

```
set(Boost_USE_STATIC_LIBS ON)
find_package(Boost COMPONENTS
    python
    REQUIRED
)
```

Using the library should then be just a matter of calling:

```
target_link_libraries(${PROJECT_NAME} PUBLIC Boost::python)
```

 - Same thing for ZMQ?

Tell me if you are interested in these changes.

tiagocoutinho replied:
> Is there a reason to stick with CMake 3.0.2 (5 years old)?

To be honest I don't know. I think only @NexeyaSGara can answer this question

> Why not use find_package() to find Boost?

There are more variables at stake:
* boost needs to be statically compiled, 
* for the correct architecture (32/64) (i guess cmake handles this one)
* use the selected python version

I don't know if this can be done "automatically" in CMake.

> set(Boost_USE_STATIC_LIBS ON)
find_package(Boost COMPONENTS
    python
    REQUIRED
)

how does cmake find boost anyway (sorry Cmake is not my strong :-P)

> Same thing for ZMQ?
Actually I don't think it should be necessary to add omniorb and zeromq at all. The only direct dependencies of pytango are boost-python and tango.

> Tell me if you are interested in these changes.

Currently, my main focus is just to be able to bring the work of @NexeyaSGara to the 9.3.1.
There is always room for improvement. I just didn't want you to loose a lot of time with this specially since the unit tests cannot run on windows yet.
I propose that if you are interested to open another PR to make these improvements after otherwise the work will never end :-)

sdebionne replied:
>I propose that if you are interested to open another PR to make these improvements after otherwise the work will never end :-)

That is what I had in mind, no showstopper here, just possible improvements on the CMake side.

> set(Boost_USE_STATIC_LIBS ON)

This tells CMake to use the static variant of the boost libraries. 

> how does cmake find boost anyway

`find_package()` look for packages in the usual places which is obviously platform specific. If not found, it declares CMake variables that the user can fill in.

In the same vein, the python interpreter, library, numpy, the target site_packages folder... can all be found automatically.

tiagocoutinho replied:
I thought `LIBRARY_VERSION` was a special variable being used by CMake.
It seems it is not serving any purpose here since the single source of truth for
the PyTango version is in `tango/release.py`. Therefore I propose to simply skip this definition completely

ajoubertza commented on appveyor.yml:
```diff
@@ -0,0 +1,324 @@
+version: 1.0.{build}
+branches:
+  only:
+    - NexeyaSGara-windows-ci
```
This needs to be updated before merging.

For testing your branch, probably needs to be `windows`.

Obviously only useful to make a final change once the PR is about to be merged.  Would it be `develop`?  Ideally we want releases to be built from tags, in which case the branch doesn't matter.   I'm not sure how that would be done in appveyor.

Having it run from `develop` would also be useful, just to see if things are still working or breaking.

tiagocoutinho replied:
Actually before I invest some time with appveyor I wanted to discuss the practicality of it.
Probably this is not the place for it. I think it should be brought up to a tango kernel meeting.

I will write some advantage/disadvantage table.

sdebionne replied:
@tiagocoutinho If you are looking for AppVeyor alternatives, look at [Azure pipelines](https://azure.microsoft.com/en-us/services/devops/pipelines/), I know projects that CI happily with MS infra.

tiagocoutinho replied:
Thanks @sdebionne. I will have a look

tiagocoutinho replied:
I propse to focus just on windows build on this PR.

Appveyor / Azure devops can come in a future PR.

ajoubertza replied:
How will you do this build, if not with AppVeyor (or Azure CI)?  We need compilation with a whole lot of different versions of Visual Studio.  And ideally in a repeatable environment, so not on your computer.  AppVeyor is already being used for cppTango.   And this PR already includes a configured `appveyor.yml` file, so not much work required at all.

tiagocoutinho replied:
> And ideally in a repeatable environment, so not on your computer

I guess in appveyor they keep windows up to date (.NET SDK, VS patches, etc) so thinking the environment is repeatable is, IMHO, an illusion. Same goes for travis. I remember not long ago you were fighting because travis decided to change the base image, right?


> AppVeyor is already being used for cppTango. And this PR already includes a configured appveyor.yml file, so not much work required at all so we need to either start a boost appveyor  project ourselves to do boost 

Sorry but I don't agree. #276 has been stalled for ages because appveyor is not building anymore and nobody has picked this up.
The boost dependencies are not available any more. Do you have an idea how we could do this? I need some help/advice here.
 
I am not saying we should give up on CI. Just saying that we should limit the scope of this PR so it doesn't drag like #276.

bourtemb replied:
Boost dependencies should be available in https://github.com/tango-controls/boost-ci/releases page.
The zip files listed there used to be appveyor artifacts.
I've done the same with the latest [cppTango release](https://github.com/tango-controls/cppTango/releases).

bourtemb replied:
In cppTango, we have similar dependencies with zmq for instance.
We simply download the corresponding zip files from the [zmq-windows-ci repo release page](https://github.com/tango-controls/zmq-windows-ci/releases
).
Here is the line on appveyor.yml downloading the zip file:
https://github.com/tango-controls/cppTango/blob/tango-9-lts/appveyor.yml#L193

tiagocoutinho replied:
Uau! didn't know this existed.

Thanks for unblocking this @bourtemb. That is the kind of push I needed

ajoubertza commented on CMakeLists.txt:
```diff
@@ -0,0 +1,254 @@
+
+#convenient versions
+set(LIBRARY_VERSION "${MAJOR_VERSION}.${MINOR_VERSION}.${PATCH_VERSION}")
+if(NOT WIN32)
```
This file is being added specifically for Windows builds, so is it really useful for other platforms?  If not useful, I suggest we just abort at this stage:
```
if(NOT WIN32)
message(FATAL_ERROR "This file is only targeting Windows builds - aborting!")
endif(NOT WIN32)
```

tiagocoutinho replied:
Agree

sdebionne replied:
Are you saying that the whole CMake effort was only for supporting Windows? I wonder what is missing to support Linux...

tiagocoutinho replied:
On linux we don't need cmake. A simple `python setup.py build` works well.

I propose that if you are interested to open another PR to make these improvements after otherwise the work will never end :-)

To be completely honest I don't think we need cmake on windows neither.

ajoubertza replied:
@tiagocoutinho I still think this needs to be modified to abort if not Windows so that people don't try to use it on other platforms.  E.g.
```
if(NOT WIN32)
message(FATAL_ERROR "This file is only targeting Windows builds - aborting!")
endif(NOT WIN32)
```

ajoubertza commented on appveyor.yml:
```diff
@@ -0,0 +1,324 @@
+  - cmd: cd "C:\projects\"
+  - cmd: md libtango
+  - cmd: cd "C:\projects\" 
+  - if %ARCH_PY%==win32-msvc9-py27 appveyor DownloadFile https://ci.appveyor.com/api/buildjobs/fibcw0rnx1ahit1d/artifacts/libtango_9.2.6_v90_x86.zip
```
These artefacts are a problem (and one of the reasons I said the previous PR would take long to complete).
They don't exist anymore.  AppVeyor only keeps them for a while before deleting.  @NexeyaSGara's idea was to store them in github files linked to releases, e.g. https://github.com/tango-controls/omniorb-windows-ci/releases.  I also see [boost-ci](https://github.com/tango-controls/boost-ci) and [zmq-windows-ci](https://github.com/tango-controls/zmq-windows-ci) repos.
I see we don't have that for https://github.com/tango-controls/cppTango/releases - it just points to AppVeyor.

Besides missing files, we would want to use at least 9.3.3, not 9.2.6.

tiagocoutinho replied:
Deferred until appveyor decision is taken (see above)

tiagocoutinho replied:
I propse to focus just on windows build on this PR.

Appveyor / Azure devops can come in a future PR.

ajoubertza replied:
I don't see how you can defer this?  The compilation depends on these artefacts existing.

ajoubertza commented on appveyor.yml:
```diff
@@ -0,0 +1,324 @@
+  #- cmd: if %ARCH%==x64-msvc12 %PYTHONPATH%python.exe winsetup.py build dist tmp release x64
+  #- cmd: if %ARCH%==x64-msvc14 %PYTHONPATH%python.exe winsetup.py build dist tmp release x64
+  #copying artifacts
+  - cmd: if %ARCH_PY%==win32-msvc9-py27 move .\dist\pytango-9.2.3.win32-py2.7.exe C:/projects/pytango
```
pytango-9.2.3 isn't right.  Another thing to parameterise/get from git tag or setup.py?

tiagocoutinho replied:
Deferred until appveyor decision is taken (see above)

tiagocoutinho replied:
I propse to focus just on windows build on this PR.

Appveyor / Azure devops can come in a future PR.

ajoubertza commented on ext/server/device_impl.cpp:
```diff
@@ -673,7 +673,12 @@ namespace PyDeviceImpl
-        Tango::Command *cmd_ptr = nullptr;
+	    #if _MSC_VER < 1600 //MSVC version <8
+	    	#define NULL ((void *)0)
+        	Tango::Command *cmd_ptr = (struct Tango::Command *) malloc(sizeof(struct Tango::Command));
```
This is weird.  Why would we need to allocate memory for one special version of compiler?  Also, the definition of `NULL` on the previous line hasn't been used.

tiagocoutinho replied:
Agree. Looks like a memory leak to me.
Do you remember why you did this @NexeyaSGara ?

ajoubertza commented on ext/server/device_impl.cpp:
```diff
@@ -1054,7 +1059,7 @@ Tango::ConstDevString Device_3ImplWrap::dev_status()
         if (override dev_status = this->get_override("dev_status") )
 	{
-            std::string status = dev_status();
+            std::string status = boost::python::call<std::string>(this->get_override("dev_status")());
```
Is the assignment to `dev_status` on line 1060 still required?  Not sure how `override` works in this context - is it just a type?
Could we do `boost::python::call<std::string>(dev_status())`

tiagocoutinho replied:
Seems that this change is not correct. Travis is failing due to this I believe.
Do you remember why you did this @NexeyaSGara ?

tiagocoutinho replied:
Should be ok now

tiagocoutinho replied:
Should be ok now

ajoubertza reviewed (changes requested):
Thanks for picking this up @tiagocoutinho.

tiagocoutinho committed:
Remove unused __prepare_nt() from __init__

sdebionne committed:
Set version from git tags.

tiagocoutinho committed:
Fix dev_status on all platforms

tiagocoutinho committed:
Resolve suspicious code (possible memory leak)

tiagocoutinho committed:
Refactor appveyor builds

- update boost to 1.70
- update cpptango to 9.3.3
- remove unused code
- add python 3.6 64bits
- add python 3.7 32 and 64 bits
- parallelize build
- produce wheels
- don't produce msi

tiagocoutinho committed:
Avoid buggy gevent

tiagocoutinho committed:
Bump version to 9.3.1dev1

tiagocoutinho committed:
Remove unused vs solution files

ajoubertza reviewed (changes requested):
@tiagocoutinho Great, I see you got a [successful appveyor build](https://ci.appveyor.com/project/tiagocoutinho/pytango/builds/24649986)! I still have some minor notes, but this is looking good.

tiagocoutinho committed:
Cleanup CMakeLists

- bail out if not on windows
- clean all WIN32 references
- better indentation

tiagocoutinho committed:
CMake and appveyor improvements

- simplify CMakeLists
- PYTHONPATH is misleading: replace with PYTHON_ROOT
- TANGO_PATH replaced with TANGO_ROOT

tiagocoutinho committed:
Merge branch 'sdebionne-windows' into windows

# Conflicts:
#	CMakeLists.txt

tiagocoutinho commented:
> @tiagocoutinho Great, I see you got a [successful appveyor build](https://ci.appveyor.com/project/tiagocoutinho/pytango/builds/24649986)! I still have some minor notes, but this is looking good.

Yes, finally something interesting!

Couple TODOs:
* [x] Handle CMakefiles PACKAGE_VERSION (thanks to @sdebionne!)
* [x] Stop CMake when not in windows
* [x] Remove unused `__init__.py:prepare_nt()`
* [x] Add Appveyor badge to README.md
* [ ] After PR is accepted change Appveyor settings:
  * [ ] PyTango github team should have full access to the appveyor project
  * [ ] default branch to build from `windows` to `develop`
  * [ ] git should point to **tango-controls/pytango instead of *tiagocoutinho/pytango*

tiagocoutinho committed:
Add appveyor build status badge to README

ajoubertza commented:
Looks good.  Please add a badge to the readme linking to the build.
(Oh, too late - you already did that!)

ajoubertza commented:
@tiagocoutinho Next question is if we can test that the these builds actually work on a Windows machine?  We aren't running the unit tests against the Windows builds.

andygotz commented:
I could try to test it on Windows. Where can I download the builds from?

ajoubertza reviewed (approved):
Looks good!  Thanks for all the work @tiagocoutinho, @sdebionne, and @NexeyaSGara. 
(Would like to have some form of testing to know that these packages actually work).

tiagocoutinho commented:
> @tiagocoutinho Next question is if we can test that the these builds actually work on a Windows machine? We aren't running the unit tests against the Windows builds.

Yes, I that would be nice. I had a try and the tests don't run. Right now they depend on linux specific python calls (like `os.fork`).
I propose to create another issue to address this.

> I could try to test it on Windows. Where can I download the builds from?

I uploaded a 9.3.1.dev1 to pypi and also to my conda channel, so from a windows machine either:
1. outside conda environment:
```bash
$ pip install gevent six  # this is just to prevent installing a buggy pre-release of gevent
$ pip install --pre pytango
$ python -c 'import tango; print(tango.__version_info__)'
(9, 3, 1, 'dev', 1)
```
2. from conda environment (python 2.7, 3.6 or 3.7):

```bash
conda install pytango -c tiagocoutinho
$ python -c 'import tango; print(tango.__version_info__)'
(9, 3, 1, 'dev', 1)
```

tiagocoutinho commented:
Is it ok to merge? If nobody disagrees I can do it

ajoubertza commented:
@tiagocoutinho I don't think it is reasonable to merge this until someone has at least run a basic server and client on a Windows computer and confirmed that these builds are at all useful.

@andygotz offered to try it out, so can we wait until he's done a test?

andygotz commented:
Yes promise stills holds. I will make time for it today. I also asked @NexeyaSGara to take a look.

andygotz commented:
OK I did it - tested the client and server part with conda on Windows 10 with Python 3.6. Both client and server examples from the PyTango documentation work - well done! 

Only issue I had was I had to update numpy and six on my laptop.

Looks good to me!

ajoubertza commented:
Thanks for testing @andygotz.  Good to merge @tiagocoutinho - you can click the button!

tiagocoutinho commented:
Thanks @andygotz. I also ran a simple python server and client and seems to work.
I will merge then.

Thanks @ajoubertza for patiently reviewing this and @sdebionne for the help with CMake
